### DISCUSSION TITLE
Support zigbee/thread groups in HA

### DISCUSSION THREAD

**Discussion by LukeDefeo** [home-assistant/architecture#882] [Entity Models]
## Problem

Home assistant treats zigbee groups whether they are from ZHA or deconz as light entities. The light entity only has turn_on and turn_off commands. This is a big pain point for groups when you want to adjust the settings for only the lights **that are on in the group**. 

There are a lot of situations where this is needed. One example: I have the side lights in a room in a zigbee group. I also have an automation to turn off one of these lights when the tv is on to avoid glare. 

If I try to adjust the brightness of this light group in home assistant, it will send a turn_on command to the light entity (which is actually a zigbee group). This will change the brightness of the lights but **also turn on all lights in that group that were off and should remain off**

## Existing proposals

A solution to the same problem was proposed [here](https://github.com/home-assistant/architecture/discussions/537), It seems to be rejected or at least in limbo. From what I can tell the main reason was due to ambiguous / undefined behaviour for what adjust should do when the lights are off, especially if the light entity represents a single light.

I am offering this as another possible solution to what is basically the same problem. That said I feel that #537 has merit and would suffice for solving this problem

### New Entity - LightGroup
Introduce new entity called LightGroup (see alternative names at the end). It is a subclass of Light, the turn off and turn on commands are not overridden.

Light group offers one new service / method called `setAttributes`. It takes the same attributes that light_on takes. It can pass these down to the underlying zigbee/matter group but without invoking the on command at the underlying layer, and voila we can adjust groups without turning them on.

By introducing a separate subclass entity we are removing the ambiguity around how setAttributes behaves for individual lights as in #537. setAttributes only really makes sense for zigbee light groups therefore the class hierarchy could reflect this. The documentation of this new service call might read as:

> Adjust attributes of this light group, This will not turn on lights in the underlying group that are currently off

### Integration points
Off the top of my head the following parts of the Home assistant platform would need to start using this

#### Lovelace slider
<img width="532" alt="image" src="https://user-images.githubusercontent.com/2492575/224821452-fbb87a37-5af1-4432-8535-56fa74a97b67.png">

This component would use setAttributes when adjusting color temperature or brightness of a light group

#### Integration - Voice assistant
When using intent setBrightness of x then it would use setAttribute if x is a light group

### Thread / matter

Im pretty confident this would apply to matter / thread since it is a close cousin of zigbee

### ZWave

I don't have any zwave gear but it looks like they also have a similar concept of groups. I don't know enough about wave to say if this concept would work for wave. If not then the wave integration would just not expose LightGroup entities.

### Migration

This is an aspect I'm not sure about since I'm not familiar enough with how entities are stored in the home assistant db. It is a solvable problem in theory but I don't know how this would work in practice, any input would be appreciated.

### Alternative names

BroadcastLightGroup - This reflects the broadcast nature of the underlying zigbee / matter / deconz layer, though it is more verbose and perhaps too technical

ExternalLightGroup - This name makes it clear that the light group member ship is managed externally to home assistant. Its a little more verbose but is probably clearest name

**Comment by balloob** [comment]
It is too confusing to have certain entities be treated differently. In Home Assistant, you send a command to a group, all entities in that group get turned on and have the command applied. There is no such thing as set state. We will never implement an exception in behavior for a single platform, as that undermines our abstraction layer and makes it unpredictable for the user.

**Reply by LukeDefeo** [reply]
@balloob @frenck 

> In Home Assistant, you send a command to a group, all entities in that group get turned on and have the command applied. 

Im not sure what you mean by group in this context - home assistant or zigbee? 

>  There is no such thing as set state.

I am not suggesting we set the home assistant state directly. Perhaps the name `setAttributes` was misleading. What if it was called `adjust` like in the the other proposal. 

The point is we need a service call that can adjust light group parameters without it being tied to the turn on command. It is still a command that goes to the underlying layer and the underlying layer will push out new state for home assistant to apply to its internal model. Nothing changes to the home assistant data flow in this regard. 

> We will never implement an exception in behaviour for a single platform, as that undermines our abstraction layer and makes it unpredictable for the user.

Im not sure this is an exception. Multiple light platforms have this functionality of groupings where a single message on the network can control many lights. The fact that home assistant doesn't model this well is a blind spot. Additionally a light group is a subclass of light so turn_on, turn_off etc work the same as before. There is no change in the abstraction if people was to carry on using it as they do today. We are just adding additionally functionality available to the hardware.

As I said Im not precious on the solution but perhaps we could first agree this is an issue that is worth solving and then together we find a solution that works for home assistant. When my partner goes to change the brightness of the side lights in home assistant and it turns on some lights I cannot tell her about abstraction layers and such.

**Reply by Gaff** [reply]
> There is no such thing as set state.

I dispute this - most entities support the "toggle" command which implicitly uses the entity state! ;)

**Comment by Gaff** [comment]
I went as far as writing a PR for just this solution - it adds an `adjustment` command to light groups which will only be dispatched to alread-on members: https://github.com/home-assistant/core/pull/78750 - it's no more stateful than the "toggle" command, which is widely supported.

Alas I couldn't get any traction for it - but if @balloob  or @frenck have time to critique the idea that would be great :D

**Reply by LukeDefeo** [reply]
The maintainers of the project seem unwilling to engage with this proposal or similar proposals.  I can only assume they don't use zigbee thread or similar devices themselves else they would see how misaligned home assistant is with zigbee / thread. Home assistant groups are *not* a solution

**Comment by Quentame** [comment]
You just need a new service to adjust light properties only if already on.

As proposed here: #537
The issue is that the service should act on lights independently and maybe not the group, as if one light is on the group is on? The implementation of the adjustments of off lights may depends on integration (ZHA, Hue ...), or use HA groups

**Comment by alive1** [comment]
I understand the technical arguments for why it's not in the interest of the home assistant software platform to introduce a "only adjust the lights of the group that are already on" but from a user perspective, this is the expected, natural and intuitive mode of operation. 

If there exists some other straightforward way of achieving this same desired effect, I'd be happy to know. So far the solutions I've seen are abusing jinja2 to achieve this basic effect. I think the effect should be achievable by means of ClickOps.

**Reply by parautenbach** [reply]
> If there exists some other straightforward way of achieving this same desired effect, I'd be happy to know.

One idea: You can dynamically create a group. Every time a light state changes to on or off, trigger an automation that updates this group.

**Comment by patienttruth** [comment]
I've read through this and #537 as well as the PR that went stale. It seems that maybe the idea fizzled with the whole adjusting the light when it's off thing.

What I came looking for was the ability to adjust only the lights in a group that are already on. If there's any chance of reviving that idea, is be stoked.
